ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -0,0 +1,256 @@
+    Iterator<Map.Entry<Integer, GetRequestInfo>> inFlightRequestsIterator =
+        correlationIdToGetRequestInfo.entrySet().iterator();
+    while (inFlightRequestsIterator.hasNext()) {
+      Map.Entry<Integer, GetRequestInfo> entry = inFlightRequestsIterator.next();
+      if (time.milliseconds() - entry.getValue().startTimeMs > routerConfig.routerRequestTimeoutMs) {
+        operationTracker.onResponse(entry.getValue().replicaId, false);
+        setOperationException(
+            new RouterException("Timed out waiting for responses", RouterErrorCode.OperationTimedOut));
+        inFlightRequestsIterator.remove();
+      } else {
+        // the entries are ordered by correlation id and time. Break on the first request that has not timed out.
+        break;
+      }
+    }
+  }
+
+  /**
+   * For this operation, create and populate get requests (in the form of {@link RequestInfo}) to send out.
+   * @param requestRegistrationCallback the {@link RequestRegistrationCallback} to call for every request that gets
+   *                                    created as part of this poll operation.
+   */
+  @Override
+  void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
+    //First, check if any of the existing requests have timed out.
+    cleanupExpiredInFlightRequests();
+
+    checkAndMaybeComplete();
+    if (isOperationComplete()) {
+      return;
+    }
+
+    Iterator<ReplicaId> replicaIterator = operationTracker.getReplicaIterator();
+    while (replicaIterator.hasNext()) {
+      ReplicaId replicaId = replicaIterator.next();
+      String hostname = replicaId.getDataNodeId().getHostname();
+      Port port = new Port(replicaId.getDataNodeId().getPort(), PortType.PLAINTEXT);
+      GetRequest getRequest = createGetRequest(blobId, getOperationFlag());
+      RequestInfo request = new RequestInfo(hostname, port, getRequest);
+      int correlationId = getRequest.getCorrelationId();
+      correlationIdToGetRequestInfo.put(correlationId, new GetRequestInfo(replicaId, time.milliseconds()));
+      requestRegistrationCallback.registerRequestToSend(this, request);
+      replicaIterator.remove();
+    }
+  }
+
+  /**
+   * Handle the given {@link ResponseInfo} and update the status of the operation.
+   * Based on the status of the response, this involves handling the body, notifying the operation tracker,
+   * notifying the response handler and possibly setting the operationException.
+   * @param responseInfo the {@link ResponseInfo} to be handled.
+   */
+  @Override
+  void handleResponse(ResponseInfo responseInfo) {
+    if (isOperationComplete()) {
+      return;
+    }
+    int correlationId = ((GetRequest) responseInfo.getRequest()).getCorrelationId();
+    // Get the GetOperation that generated the request.
+    GetRequestInfo getRequestInfo = correlationIdToGetRequestInfo.remove(correlationId);
+    if (getRequestInfo == null) {
+      // Ignore right away. This associated operation has completed.
```

Comment:
That would be hard, and is not really necessary. Ultimately there should be a response for every request sent out to the `NetworkClient`, either because it times out at the `NetworkClient` itself, or because the connection on which it sent out a request disconnects, or whenever the server responds.

That said, I noticed a bug in the `GetManager#handleResponse()` where it should remove the entry from `correlationIdToGetRequestInfo`. Will fix that.